### PR TITLE
Stop monitoring goroutine when removing replica

### PR DESCRIPTION
**Pull request by keithalucas**
The monitoring goroutine was continuing to run after the replica was
removed.  This caused a problem when a replica is recreated for the same
IP address and port after one is deleted.  When a replica is removed,
the channel that the monitoring goroutine is waiting for will eventually
get an error message.  In the normal case, the SetReplicaMode won't find
a replica with the address to mark as ERR.  However, when another
replica is with the same address is added after the original was
removed, SetReplicaMode for the old monitorChan will mark the new
replica as ERR with SetReplicaMode.

This change makes removing the replica also stop the monitoring
goroutine that was launched for that replica.  It puts a nil error on
the replica's monitor channel so monitoring goroutine  will exit as part
of removing the replica.

longhorn/longhorn#1628

Signed-off-by: Keith Lucas <keith.lucas@suse.com>

### CODE REVIEW

File: pkg/controller/replicator.go
Comment (joshimoo) [inline pkg/controller/replicator.go:80]: The comment below mentions that backend can be nil, then this would throw a nil pointer.
Did you investigate whether it's safe to rely on this not being nil?

**Review by joshimoo — CHANGES_REQUESTED** [review]
the below comment mentions that backend can be nil, is this a concern?

File: pkg/controller/replicator.go
Comment (joshimoo) [inline pkg/controller/replicator.go:81]: Instead of directly writing against the monitoring channel we should instead consider calling.
backend.backend.StopMonitoring()

Since the above, will lead to a deadlock, for file based backends which return a nil channel.

**Review by joshimoo — CHANGES_REQUESTED** [review]

File: pkg/controller/replicator.go
Comment (keithalucas) [reply pkg/controller/replicator.go]: I moved it into the if.

File: pkg/controller/replicator.go
Comment (keithalucas) [reply pkg/controller/replicator.go]: I changed it to StopMonitoring.

**Review by joshimoo — APPROVED** [review]
LGTM

I think stopping monitoring is a good change I am not sure if we are covering all cases. Since I am not sure when backend.backend would be nil and if that case also handles monitoring cancellation correctly.

But this is already an improvement, so we could merge it already.
As quickly discussed during the call one refactoring that we could do is replace the closeChannel with a context.WithCancel then just call the cancel function during stop monitoring/close.

**Comment by joshimoo** [comment]
@keithalucas I approved it, if you feel comfortable with the implementation you can use the `rebase and merge` button or let me know and I will do it instead :) I saw there was some issue with reproduction during the team meeting, just double checking if this impacts this pr?

**Comment by joshimoo** [comment]
Only maintainers can merge even after approval, so I will merge it. Thanks for reminding me @keithalucas

**Pull request merged**
